### ISSUE TITLE
problem with tophat

### ISSUE THREAD

**Issue by mra8187** [bioconda/bioconda-recipes#2752]
tophat2 -p 4 -G '/home/mra/DATA/Anotation.gtf' -o '/home/mra/DATA/out' ref '/home/mra/DATA/genome_index_base.fa' '/home/mra/DATA/1.fastq.gz' '/home/mra/DATA/2.fastq.gz' 

[2016-10-31 20:21:41] Beginning TopHat run (v2.1.1)
-----------------------------------------------
Traceback (most recent call last):
  File "/home/mra/miniconda3/bin/tophat", line 4107, in <module>
    sys.exit(main())
  File "/home/mra/miniconda3/bin/tophat", line 3899, in main
    run_log = open(logging_dir + "run.log", "w", 0)
ValueError: can't have unbuffered text I/O

**Comment by mra8187** [comment]
tophat2 -p 4 -G '/home/mra/DATA/Anotation.gtf' -o '/home/mra/DATA/out' ref '/home/mra/DATA/genome_index_base.fa' '/home/mra/DATA/1.fastq.gz' '/home/mra/DATA/2.fastq.gz' 

[2016-10-31 20:21:41] Beginning TopHat run (v2.1.1)
-----------------------------------------------
Traceback (most recent call last):
  File "/home/mra/miniconda3/bin/tophat", line 4107, in <module>
    sys.exit(main())
  File "/home/mra/miniconda3/bin/tophat", line 3899, in main
    run_log = open(logging_dir + "run.log", "w", 0)
ValueError: can't have unbuffered text I/O

**Comment by daler** [comment]
This error is from running TopHat using Python 3.

Tophat itself is not supported under python3. We tried to fix that in the bioconda package using the 2to3.py script, but it appears that conversion was incomplete and will require tests to make sure it works.

The immediate fix would be to run tophat in a python2 environment:

```bash
conda create -n tophat-env python=2 tophat
```

@bioconda/core, I think one solution would be to drop python3 support for tophat; another would be to fix the conversion but it would need more extensive tests (e.g., example fasta and fastq files) in order to ensure the py3 compatibility fixes worked. Any preferences? 

@bgruening maybe a hackathon task could be a tophat test for python3.

**Comment by bgruening** [comment]
Like the idea with the hackathon task, otherwise it should be ok to drop python3 support.

**Comment by mra8187** [comment]
thank you very much ..... i work on it for 2 day at least . and i never can
fix it.....  thank you

**Comment by mra8187** [comment]
dear all
when i'm using tophat that install by conda , why the tophat not cheacking
the samtools ?
may this problem cause the low percentage of mapping ?
because my data have good quality and in any way my mapping percentage is
about 50 % .....

thanks

**Comment by daler** [comment]
bioconda tries to make tools and dependencies easy to *install*. We are unable to provide support for *using* all the software we make available. If tophat installed and is at least working, you may have better luck on biostars.org or on the authors [google group](https://groups.google.com/forum/#!forum/tuxedo-tools-users).

**Comment by mra8187** [comment]
thanks

**Comment by mikedamour** [comment]
Having same problem as mra8187 for a month. Just found this thread.
@bioconda/core - Question about preference, for me it's worth persisting on the py3 support.  I am still confused about dual-py support on my machine.  Of course, easy for me to say.  Unfortunately, I'm not capable of making the fix myself.  Thanks for your support.  bioconda is the only way I can proceed.

**Comment by bgruening** [comment]
@mikedamour can you work on getting tophat python3 ready?

**Comment by mikedamour** [comment]
@bgruening, sorry to respond late. Unfortunately I'm not up to the task to complete the Py3 conversion, especially if the auto-conversion scripts have failed already. I have tried another alignment tool and hope to be able to complete with that. Thanks for your response.

**Comment by rvalieris** [comment]
closing this since tophat python3 packages were removed by this PR: https://github.com/bioconda/bioconda-recipes/pull/3513

tl;dr is: tophat doesn't support python3 so we can't really support it either without some serious work on it, so its best to drop the package to not create confusion.

**Comment by vsoch** [comment]
Just ran into this as well :/ +1 python 3 support!

**Comment by mikedamour** [comment]
@vsoch Vanessa Scott, found that was best to switch from tophat to hisat2. New algorithm and, apparently, lots of pre-compute of indices gives many time faster results. Saw references to 50x faster but I didn't do A/B testing. Be sure to download the pre-indexed reference genome. I found it on the Johns-Hopkins site a year ago, I think. Maybe Bioconda has provisions for it now.

**Comment by vsoch** [comment]
thanks @mikedamour I will give it a go! And please say hello to the other Vanessa! (that happens to be my name too :), and I have all the same letters  in my last name too... eerie!

**Comment by mikedamour** [comment]
@vsoch, Oops, thought I was writing your name, but bet you knew that :) Think you will like Hisat. I was using it to align huge reads files from public data. It allows you to just give the accession codes on the command line and it will download the files in parallel while it's doing alignment to the reference genome. You can then just keep the SAM files if you're space limited. As I recall, at the time it didn't have BAM output so the script did the SAM to BAM compression immediately after.
